File: hdf-converters/shared/go/cvss_test.go
package shared

import (
	"testing"

	hdf "github.com/mitre/hdf-libs/hdf-schema/dist/go/v3"
	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
)

func TestCvssVersionFromVector(t *testing.T) {
	tests := []struct {
		name   string
		vector string
		want   hdf.Version
	}{
		{"v2.0 prefix", "CVSS:2.0/AV:N/AC:L", hdf.The20},
		{"v3.0 prefix", "CVSS:3.0/AV:N/AC:L", hdf.The30},
		{"v3.1 prefix falls through to default", "CVSS:3.1/AV:N/AC:L", hdf.The31},
		{"v4.0 prefix", "CVSS:4.0/AV:N/AC:L", hdf.The40},
		{"empty defaults to 3.1", "", hdf.The31},
		{"malformed/unprefixed defaults to 3.1", "AV:N/AC:L", hdf.The31},
	}
	for _, tc := range tests {
		t.Run(tc.name, func(t *testing.T) {
			assert.Equal(t, tc.want, CvssVersionFromVector(tc.vector))
		})
	}
}

func TestCvssVersionFromString(t *testing.T) {
	tests := []struct {
		name string
		in   string
		want hdf.Version
	}{
		{"2.0", "2.0", hdf.The20},
		{"3.0", "3.0", hdf.The30},
		{"3.1 falls through to default", "3.1", hdf.The31},
		{"4.0", "4.0", hdf.The40},
		{"empty defaults to 3.1", "", hdf.The31},
		{"unrecognized defaults to 3.1", "9.9", hdf.The31},
	}
	for _, tc := range tests {
		t.Run(tc.name, func(t *testing.T) {
			assert.Equal(t, tc.want, CvssVersionFromString(tc.in))
		})
	}
}

func TestCvssSeverityFromScore(t *testing.T) {
	tests := []struct {
		score float64
		want  hdf.CVSSSeverity
	}{
		{0.0, hdf.None},
		{0.1, hdf.CVSSSeverityLow},
		{3.9, hdf.CVSSSeverityLow},
		{4.0, hdf.CVSSSeverityMedium},
		{6.9, hdf.CVSSSeverityMedium},
		{7.0, hdf.CVSSSeverityHigh},
		{8.9, hdf.CVSSSeverityHigh},
		{9.0, hdf.CVSSSeverityCritical},
		{10.0, hdf.CVSSSeverityCritical},
	}
	for _, tc := range tests {
		assert.Equal(t, tc.want, CvssSeverityFromScore(tc.score))
	}
}

func floatPtr(f float64) *float64 { return &f }

func TestBuildCvss(t *testing.T) {
	t.Run("all base fields present", func(t *testing.T) {
		cv := BuildCvss(CvssInput{
			Version:    hdf.The31,
			BaseScore:  floatPtr(9.8),
			BaseVector: "CVSS:3.1/AV:N/AC:L/PR:N/UI:N/S:U/C:H/I:H/A:H",
			Source:     "CVE-2021-44228",
		})
		assert.Equal(t, hdf.The31, cv.Version)
		require.NotNil(t, cv.BaseScore)
		assert.InDelta(t, 9.8, *cv.BaseScore, 0.001)
		require.NotNil(t, cv.BaseSeverity)
		assert.Equal(t, hdf.CVSSSeverityCritical, *cv.BaseSeverity)
		require.NotNil(t, cv.BaseVector)
		assert.Equal(t, "CVSS:3.1/AV:N/AC:L/PR:N/UI:N/S:U/C:H/I:H/A:H", *cv.BaseVector)
		require.NotNil(t, cv.Source)
		assert.Equal(t, "CVE-2021-44228", *cv.Source)
	})

	t.Run("score only, no vector or source", func(t *testing.T) {
		cv := BuildCvss(CvssInput{Version: hdf.The20, BaseScore: floatPtr(0.0)})
		assert.Equal(t, hdf.The20, cv.Version)
		require.NotNil(t, cv.BaseScore)
		assert.InDelta(t, 0.0, *cv.BaseScore, 0.001)
		require.NotNil(t, cv.BaseSeverity)
		assert.Equal(t, hdf.None, *cv.BaseSeverity)
		assert.Nil(t, cv.BaseVector)
		assert.Nil(t, cv.Source)
	})

	t.Run("vector only, no score", func(t *testing.T) {
		cv := BuildCvss(CvssInput{Version: hdf.The40, BaseVector: "CVSS:4.0/AV:N/AC:L"})
		assert.Equal(t, hdf.The40, cv.Version)
		assert.Nil(t, cv.BaseScore)
		assert.Nil(t, cv.BaseSeverity)
		require.NotNil(t, cv.BaseVector)
		assert.Equal(t, "CVSS:4.0/AV:N/AC:L", *cv.BaseVector)
		assert.Nil(t, cv.Source)
	})

	t.Run("version only, all optional fields omitted", func(t *testing.T) {
		cv := BuildCvss(CvssInput{Version: hdf.The31})
		assert.Equal(t, hdf.The31, cv.Version)
		assert.Nil(t, cv.BaseScore)
		assert.Nil(t, cv.BaseSeverity)
		assert.Nil(t, cv.BaseVector)
		assert.Nil(t, cv.Source)
	})
}

File: hdf-converters/shared/typescript/cvss.test.ts
import { describe, it, expect } from 'vitest';
import { CVSSSeverity, Version as CvssVersion } from '@mitre/hdf-schema';
import { cvssVersionFromVector, cvssVersionFromString, cvssSeverityFromScore, buildCvss } from './cvss.js';

describe('cvssVersionFromVector', () => {
  it('maps recognized prefixes to the schema Version enum', () => {
    expect(cvssVersionFromVector('CVSS:2.0/AV:N/AC:L')).toBe(CvssVersion.The20);
    expect(cvssVersionFromVector('CVSS:3.0/AV:N/AC:L')).toBe(CvssVersion.The30);
    expect(cvssVersionFromVector('CVSS:4.0/AV:N/AC:L')).toBe(CvssVersion.The40);
  });

  it('defaults to 3.1 for the 3.1 prefix, empty, undefined, and unprefixed vectors', () => {
    expect(cvssVersionFromVector('CVSS:3.1/AV:N/AC:L')).toBe(CvssVersion.The31);
    expect(cvssVersionFromVector('')).toBe(CvssVersion.The31);
    expect(cvssVersionFromVector(undefined)).toBe(CvssVersion.The31);
    expect(cvssVersionFromVector('AV:N/AC:L')).toBe(CvssVersion.The31);
  });

  it('uses a caller-supplied default for absent/unrecognized prefixes; a recognized prefix wins', () => {
    // Nessus historical output defaults to 3.0.
    expect(cvssVersionFromVector('', CvssVersion.The30)).toBe(CvssVersion.The30);
    expect(cvssVersionFromVector('AV:N/AC:L', CvssVersion.The30)).toBe(CvssVersion.The30);
    expect(cvssVersionFromVector('CVSS:3.1/AV:N', CvssVersion.The30)).toBe(CvssVersion.The31);
    expect(cvssVersionFromVector('CVSS:3.0/AV:N', CvssVersion.The30)).toBe(CvssVersion.The30);
  });
});

describe('cvssVersionFromString', () => {
  it('maps recognized version numbers to the schema Version enum', () => {
    expect(cvssVersionFromString('2.0')).toBe(CvssVersion.The20);
    expect(cvssVersionFromString('3.0')).toBe(CvssVersion.The30);
    expect(cvssVersionFromString('4.0')).toBe(CvssVersion.The40);
  });

  it('defaults to 3.1 for 3.1, empty, undefined, and unrecognized values', () => {
    expect(cvssVersionFromString('3.1')).toBe(CvssVersion.The31);
    expect(cvssVersionFromString('')).toBe(CvssVersion.The31);
    expect(cvssVersionFromString(undefined)).toBe(CvssVersion.The31);
    expect(cvssVersionFromString('9.9')).toBe(CvssVersion.The31);
  });
});

describe('cvssSeverityFromScore', () => {
  it('maps scores to the correct band across every threshold', () => {
    expect(cvssSeverityFromScore(0.0)).toBe(CVSSSeverity.None);
    expect(cvssSeverityFromScore(0.1)).toBe(CVSSSeverity.Low);
    expect(cvssSeverityFromScore(3.9)).toBe(CVSSSeverity.Low);
    expect(cvssSeverityFromScore(4.0)).toBe(CVSSSeverity.Medium);
    expect(cvssSeverityFromScore(6.9)).toBe(CVSSSeverity.Medium);
    expect(cvssSeverityFromScore(7.0)).toBe(CVSSSeverity.High);
    expect(cvssSeverityFromScore(8.9)).toBe(CVSSSeverity.High);
    expect(cvssSeverityFromScore(9.0)).toBe(CVSSSeverity.Critical);
    expect(cvssSeverityFromScore(10.0)).toBe(CVSSSeverity.Critical);
  });
});

describe('buildCvss', () => {
  it('assembles all base fields when present', () => {
    const cv = buildCvss({
      version: CvssVersion.The31,
      baseScore: 9.8,
      baseVector: 'CVSS:3.1/AV:N/AC:L/PR:N/UI:N/S:U/C:H/I:H/A:H',
      source: 'CVE-2021-44228',
    });
    expect(cv.version).toBe(CvssVersion.The31);
    expect(cv.baseScore).toBe(9.8);
    expect(cv.baseSeverity).toBe(CVSSSeverity.Critical);
    expect(cv.baseVector).toBe('CVSS:3.1/AV:N/AC:L/PR:N/UI:N/S:U/C:H/I:H/A:H');
    expect(cv.source).toBe('CVE-2021-44228');
  });

  it('emits a 0.0 baseScore (with none severity) but omits an empty vector/source', () => {
    const cv = buildCvss({version: CvssVersion.The20, baseScore: 0.0});
    expect(cv.version).toBe(CvssVersion.The20);
    expect(cv.baseScore).toBe(0.0);
    expect(cv.baseSeverity).toBe(CVSSSeverity.None);
    expect(cv.baseVector).toBeUndefined();
    expect(cv.source).toBeUndefined();
  });

  it('omits baseScore/baseSeverity when the score is absent or non-finite', () => {
    const nullScore = buildCvss({version: CvssVersion.The40, baseVector: 'CVSS:4.0/AV:N/AC:L', baseScore: null});
    expect(nullScore.baseScore).toBeUndefined();
    expect(nullScore.baseSeverity).toBeUndefined();
    expect(nullScore.baseVector).toBe('CVSS:4.0/AV:N/AC:L');

    const nanScore = buildCvss({version: CvssVersion.The31, baseScore: Number.NaN});
    expect(nanScore.baseScore).toBeUndefined();
    expect(nanScore.baseSeverity).toBeUndefined();
  });

  it('omits empty baseVector and source (empty string is falsy)', () => {
    const cv = buildCvss({version: CvssVersion.The31, baseScore: 5.0, baseVector: '', source: ''});
    expect(cv.baseVector).toBeUndefined();
    expect(cv.source).toBeUndefined();
    expect(cv.baseScore).toBe(5.0);
  });

  it('emits only version when all optional fields are omitted', () => {
    const cv = buildCvss({version: CvssVersion.The31});
    expect(cv.version).toBe(CvssVersion.The31);
    expect(cv.baseScore).toBeUndefined();
    expect(cv.baseSeverity).toBeUndefined();
    expect(cv.baseVector).toBeUndefined();
    expect(cv.source).toBeUndefined();
  });
});
